File: Lectures/OS_Highlights/Week_02/race_demo.py
# race_demo.py
import threading, time, random

counter = 0


def increment():
    global counter
    for _ in range(10000):
        # Not atomic: read, increment, write
        time.sleep(random.random() * 0.0001)
        temp = counter
        temp += 1
        counter = temp


threads = [threading.Thread(target=increment) for _ in range(5)]

for t in threads:
    t.start()
for t in threads:
    t.join()

print("Expected:", 5 * 1000)
print("Actual:", counter)

File: Lectures/OS_Highlights/Week_02/race_demo.cpp
// race_demo.cpp
#include <iostream>
#include <thread>
#include <vector>

long counter = 0;

void increment() {
    for (int i = 0; i < 100000; ++i) {
        // Not atomic: read -> modify -> write
        long temp = counter;
        temp += 1;
        counter = temp;
    }
}

int main() {
    std::vector<std::thread> threads;

    for (int i = 0; i < 5; ++i)
        threads.emplace_back(increment);

    for (auto& t : threads)
        t.join();

    std::cout << "Expected: " << 5 * 100000 << "\n";
    std::cout << "Actual:   " << counter << "\n";

    return 0;
}